lib/hammox/type_engine.ex

```diff
@@ -352,6 +352,22 @@ defmodule Hammox.TypeEngine do
 
       [{:type, _, :map_field_exact, [{:atom, _, :__struct__}, {:atom, _, other_struct_name}]}] ->
         {:error, [{:struct_name_type_mismatch, struct_name, other_struct_name}]}
+
+      [{:type, _, :map_field_exact, [{:atom, _, :__struct__}, {:type, _, _, _} = struct_type]}] ->
+        case {match_type(struct_name, struct_type),
+              match_type(Map.delete(value, :__struct__), {:type, 0, :map, rest_field_types})} do
+          {:ok, :ok} ->
+            :ok
+
+          {:ok, error} ->
+            error
+
+          {error, :ok} ->
+            error
+
+          {{:error, struct_reasons}, {:error, rest_reasons}} ->
+            {:error, struct_reasons ++ rest_reasons}
+        end
```

Comment:
Hammox only ever really returns one type error reason — the "reasons" array is more like a "reason stacktrace". So we can't concat two — if you look at the error message from the `"fail with incorrect __struct__ and fields"` test, you'll see this:
```
     ** (Hammox.TypeMatchError) 
     Returned value %{__struct__: "wrong type", key: "not a number"} does not match type %{__struct__: atom(), key: number()}.
       Value "wrong type" does not match type atom().
         Map value "not a number" for key :key does not match map value type number().
           Value "not a number" does not match type integer() | float().
```
The first 2 lines are from the first error, the last 2 lines are from the second, but they are going to be formatted like one stacktrace.

We could simplify to the following:
```suggestion
        with :ok <- match_type(struct_name, struct_type),
             :ok <-
               match_type(Map.delete(value, :__struct__), {:type, 0, :map, rest_field_types}),
             do: :ok
        end
```

Reply:
Ah, interesting, OK! I mis-read some of the clauses here that used `reduce`; there was nothing quite like this case to base it on that I could find.

Change is made in 40225d2; let me know if I need to fix anything else!